# Data Analysis and Cleaning of MatchedBeer Dataset

The MatchedBeer dataset combines data from BeerAdvocate and RateBeer, two of the most prominent online platforms for beer ratings. This combined dataset was created to address the herding effect in online ratings, where early ratings can skew subsequent user opinions. By merging the two platforms, the MatchedBeer dataset allows for the study of the same beers rated independently on each platform, thereby providing a unique opportunity to analyze unbiased rating patterns.

## Matching Process Overview
The matching process is described in detail in the paper ["When Sheep Shop: Measuring Herding Effects in Product Ratings with Natural Experiments"](https://dlab.epfl.ch/people/west/pub/Lederrey-West_WWW-18.pdf), focusing on accurately aligning beers from both websites. The matching is performed in two phases to prioritize precision:

1. **Brewery Alignment**: Breweries across BeerAdvocate and RateBeer are aligned based on their exact location (state or country), with names represented as TF-IDF vectors to compute cosine similarity.
2. **Beer Alignment**: Within matched breweries, beers are aligned based on name similarity and require an exact match in alcohol by volume (ABV). Before matching, any shared tokens in brewery and beer names are removed to improve accuracy.

The algorithm ensures high precision by setting strict thresholds for cosine similarity and enforcing a significant gap between the best and second-best matches. This approach results in a dataset where brewery matching has a reported precision of 99.6%, and beer matching achieves a precision of 100%, meaning that only highly confident matches are included, though some potential matches may be excluded.

## Structure of MatchedBeer Dataset
The dataset contains essential information on beers, breweries, users, and user ratings from both platforms. Since the user bases of BeerAdvocate and RateBeer differ geographically, the combined dataset reflects a more balanced view. BeerAdvocate is more U.S.-centric, while RateBeer has a more internationally diverse user base. Despite these differences, the matched dataset has been verified for both internal and external validity, making it well-suited for analyzing the influence of initial ratings on user perceptions across platforms.

MatchedBeer provides insights into how early ratings impact subsequent ones by comparing ratings for the same beer on both sites, especially when early ratings on each platform differ significantly. By standardizing and aligning ratings across time and platforms, MatchedBeer reveals patterns that allow for the identification of herding effects, helping researchers understand the biases in user-generated ratings.

In [1]:
# Import all the necessary libraries
import polars as pl
import tqdm
import os

# Import the file in the utils folder
import sys
sys.path.append('../../utils')
from data_desc import describe_numbers, describe
from matched_dataset import load_and_process_files

# Define the paths
SRC_DATA_PATH = '../../../data/MatchedBeer'
DST_DATA_PATH = '../../../data/MatchedBeer/processed'

# Create the DST_DATA_PATH if it does not exist
if not os.path.exists(DST_DATA_PATH):
    os.makedirs(DST_DATA_PATH)

#### Data Exploration

In [2]:
file_paths = [SRC_DATA_PATH + "/beers.csv", SRC_DATA_PATH + "/breweries.csv", SRC_DATA_PATH + "/ratings.csv", SRC_DATA_PATH + "/users_approx.csv"]
dataframes_list = load_and_process_files(file_paths)

df_beers = dataframes_list[0]
df_breweries = dataframes_list[1]
df_ratings = dataframes_list[2]
df_users = dataframes_list[3]

##### Beers dataset
We'll start by looking at the beers dataset.

In [3]:
df_beers.sample(5)

shape: (5, 33)
┌────────┬────────┬────────────┬────────────┬───┬────────────┬───────────┬────────────┬────────────┐
│ abv_ba ┆ avg_ba ┆ avg_comput ┆ avg_matche ┆ … ┆ style_scor ┆ zscore_rb ┆ diff_score ┆ sim_scores │
│ ---    ┆ ---    ┆ ed_ba      ┆ d_valid_ra ┆   ┆ e_rb       ┆ ---       ┆ s          ┆ ---        │
│ f64    ┆ f64    ┆ ---        ┆ tings_ba   ┆   ┆ ---        ┆ f64       ┆ ---        ┆ f64        │
│        ┆        ┆ f64        ┆ ---        ┆   ┆ f64        ┆           ┆ f64        ┆            │
│        ┆        ┆            ┆ f64        ┆   ┆            ┆           ┆            ┆            │
╞════════╪════════╪════════════╪════════════╪═══╪════════════╪═══════════╪════════════╪════════════╡
│ 7.7    ┆ null   ┆ null       ┆ null       ┆ … ┆ null       ┆ null      ┆ 1.0        ┆ 1.0        │
│ 4.9    ┆ 3.62   ┆ 3.5575     ┆ 3.856667   ┆ … ┆ null       ┆ -0.255436 ┆ 0.723186   ┆ 1.0        │
│ 7.2    ┆ 4.01   ┆ 4.01       ┆ 4.06       ┆ … ┆ null       ┆ 0.773624  ┆ 1.0        ┆ 1.0        │
│ 5.7    ┆ 3.99   ┆ 4.01       ┆ 3.97       ┆ … ┆ null       ┆ 0.820444  ┆ 0.442664   ┆ 0.903038   │
│ 4.8    ┆ 3.25   ┆ 3.25       ┆ null       ┆ … ┆ null       ┆ -0.183587 ┆ 1.0        ┆ 1.0        │
└────────┴────────┴────────────┴────────────┴───┴────────────┴───────────┴────────────┴────────────┘

This dataset has the following structure

| Column Name | Description |
|-------------|-------------|
| `beer_id_ba` / `beer_id_rb` | Unique identifier for each beer in BA/RB | 
| `beer_name_ba` / `beer_name_rb` | Name of the beer in BA/RB |
| `beer_wout_brewery_name_ba` / `beer_wout_brewery_name_rb` | Beer name without brewery name in BA/RB |
| `brewery_id_ba` / `brewery_id_rb` | Unique identifier for the brewery in BA/RB |
| `brewery_name_ba` / `brewery_name_rb` | Name of the brewery in BA/RB |
| `style_ba` / `style_rb` | Beer style in BA/RB |
| `abv_ba` / `abv_rb` | Alcohol By Volume percentage in BA/RB |
| `nbr_ratings_ba` / `nbr_ratings_rb` | Number of ratings in BA/RB |
| `nbr_reviews_ba` | Number of reviews in BA |
| `avg_ba` / `avg_rb` | Average rating in BA/RB |
| `avg_computed_ba` / `avg_computed_rb` | Computed average rating in BA/RB |
| `zscore_ba` / `zscore_rb` | Standardized score in BA/RB |
| `nbr_matched_valid_ratings_ba` / `nbr_matched_valid_ratings_rb` | Number of matched valid ratings in BA/RB |
| `avg_matched_valid_ratings_ba` / `avg_matched_valid_ratings_rb` | Average of matched valid ratings in BA/RB |
| `ba_score_ba` | BeerAdvocate score |
| `bros_score_ba` | Bros score in BA |
| `overall_score_rb` | Overall score in RB |
| `style_score_rb` | Style-specific score in RB |
| `diff_scores` | Difference in scores between BA and RB |
| `sim_scores` | Similarity in scores between BA and RB |

Now that we have an idea what the structure of our datasets looks like, let's see if our dataset is complete or whether it contains a lot of missing entries.

In [4]:
describe(df_beers)

+------------------------------+---------+------------------+---------------+-----------+-----------------------+-------------------+
| Column                       | Type    |   Not null count |   Nulls count |   Nulls % |   Unique values count |   Unique values % |
|------------------------------+---------+------------------+---------------+-----------+-----------------------+-------------------|
| abv_ba                       | Float64 |            45640 |             0 |    0.00 % |                   351 |            0.77 % |
| avg_ba                       | Float64 |            40285 |          5355 |   11.73 % |                   373 |            0.93 % |
| avg_computed_ba              | Float64 |            40285 |          5355 |   11.73 % |                 12090 |           30.01 % |
| avg_matched_valid_ratings_ba | Float64 |            28272 |         17368 |   38.05 % |                  6658 |           23.55 % |
| ba_score_ba                  | Float64 |            10499 | 

In [5]:
describe_numbers(df_beers, filters =["beer_id_ba", "brewery_id_ba", "beer_id_rb", "brewery_id_rb"])

+------------------------------+-----------+-----------+-----------+------------+-----------+----------+---------+
| Column                       |      Mean |       Std |       25% |        50% |       75% |      Min |     Max |
|------------------------------+-----------+-----------+-----------+------------+-----------+----------+---------|
| abv_ba                       |   6.32086 |   1.85773 |         5 |          6 |       7.2 |     0.38 |    67.5 |
| avg_ba                       |   3.73237 |  0.439988 |      3.52 |       3.78 |         4 |        1 |       5 |
| avg_computed_ba              |   3.72785 |  0.424824 |      3.52 |     3.7725 |         4 |        1 |       5 |
| avg_matched_valid_ratings_ba |   3.74777 |  0.484775 |      3.52 |    3.80833 |     4.045 |        1 |       5 |
| ba_score_ba                  |    84.871 |   3.28287 |        83 |         85 |        86 |       63 |     100 |
| bros_score_ba                |   86.1623 |    8.5394 |        83 |         88 

1. Alcohol By Volume (ABV)
   - The `abv_ba` and `abv_rb` columns show identical distributions across BeerAdvocate and RateBeer, with a mean ABV of approximately 6.32%.
   - This similarity suggests a good alignment in the types of beers reviewed on both platforms. The maximum ABV of 67.5% is possibly an outlier, likely representing rare extreme beers, as typical ABVs are much lower.

2. Average Ratings (BeerAdvocate vs. RateBeer)
   - The `avg_ba` (BeerAdvocate) mean is 3.73, while `avg_rb` (RateBeer) is significantly lower at 3.14.
   - This aligns with previous findings that BeerAdvocate users tend to rate beers more generously than RateBeer users. The higher `std` value for `avg_ba` also indicates a wider range in BeerAdvocate’s ratings, suggesting more variability in user opinions.

3. Computed Averages (Consistency in Ratings)
   - `avg_computed_ba` and `avg_computed_rb` refer to recalculated averages, showing slightly lower standard deviations than their direct counterparts (`avg_ba` and `avg_rb`), indicating a slight smoothing effect.
   - Both BeerAdvocate and RateBeer have computed averages (`avg_computed_ba` vs. `avg_computed_rb`) that are fairly close to the respective raw averages, implying that the original averages align well with user trends.

4. Rating Counts
   - The median `nbr_ratings_ba` (3) and `nbr_ratings_rb` (5) are relatively low, showing that many beers have few ratings on both sites, though RateBeer has slightly more ratings per beer on average.
   - RateBeer’s broader user base also shows a higher standard deviation in the count of matched valid ratings (`nbr_matched_valid_ratings_rb` is 80.10 vs. 42.51 for BeerAdvocate), meaning some beers have far more reviews than others on RateBeer.

5. Z-scores (Standardized Ratings)
   - The mean `zscore_ba` (-0.41) is lower than `zscore_rb` (-0.10), suggesting BeerAdvocate users' ratings tend to deviate more below the mean, which may imply a slight rating inflation on RateBeer.
   - BeerAdvocate’s z-scores also exhibit a broader spread, evidenced by a standard deviation of 0.81 compared to RateBeer's 0.73, reinforcing the idea that BeerAdvocate reviews show more polarized opinions.

6. Score Difference and Similarity
   - `diff_scores` has a mean of 0.798, which indicates a considerable average difference in the ratings across the two platforms.
   - `sim_scores`, however, is high (mean of 0.986), suggesting that while individual score values differ, the rating trends between matched pairs are generally consistent.

##### Breweries dataset

Now we can take a look at our breweries dataframe. For each platform, every brewery has a location, a name, and the amount of beers they produce, along with a unique identifier. We also have the similarity and difference scores between the breweries from the two platforms. The brewery location can be very useful for our research questions. We will look into this more later on in this notebook.

In [6]:
df_breweries.sample(5)

shape: (5, 10)
┌───────┬────────────┬────────────┬────────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ id_ba ┆ location_b ┆ name_ba    ┆ nbr_beers_ ┆ … ┆ name_rb   ┆ nbr_beers ┆ diff_scor ┆ sim_score │
│ ---   ┆ a          ┆ ---        ┆ ba         ┆   ┆ ---       ┆ _rb       ┆ es        ┆ s         │
│ i64   ┆ ---        ┆ str        ┆ ---        ┆   ┆ str       ┆ ---       ┆ ---       ┆ ---       │
│       ┆ str        ┆            ┆ i64        ┆   ┆           ┆ i64       ┆ f64       ┆ f64       │
╞═══════╪════════════╪════════════╪════════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ 13444 ┆ Denmark    ┆ Heimdal-Br ┆ 0          ┆ … ┆ Heimdal-B ┆ 11        ┆ 0.345058  ┆ 1.0       │
│       ┆            ┆ yg 2       ┆            ┆   ┆ ryg       ┆           ┆           ┆           │
│ 14294 ┆ England    ┆ Newby Wyke ┆ 5          ┆ … ┆ Newby     ┆ 77        ┆ 0.764933  ┆ 0.97721   │
│       ┆            ┆ Brewery    ┆            ┆   ┆ Wyke      ┆           ┆           ┆           │
│ 42076 ┆ United     ┆ Red Clay   ┆ 8          ┆ … ┆ Red Clay  ┆ 8         ┆ 0.386143  ┆ 1.0       │
│       ┆ States,    ┆ Brewing    ┆            ┆   ┆ Brewing   ┆           ┆           ┆           │
│       ┆ Alabama    ┆ Company    ┆            ┆   ┆ Company   ┆           ┆           ┆           │
│ 42578 ┆ United     ┆ Leikam     ┆ 8          ┆ … ┆ Leikam    ┆ 8         ┆ 0.770562  ┆ 1.0       │
│       ┆ States,    ┆ Brewing    ┆            ┆   ┆ Brewing   ┆           ┆           ┆           │
│       ┆ Oregon     ┆            ┆            ┆   ┆           ┆           ┆           ┆           │
│ 2696  ┆ United     ┆ Jarre      ┆ 4          ┆ … ┆ Jarre     ┆ 9         ┆ 0.553282  ┆ 0.96798   │
│       ┆ States,    ┆ Creek      ┆            ┆   ┆ Creek     ┆           ┆           ┆           │
│       ┆ Colorado   ┆ Ranch      ┆            ┆   ┆ Ranch     ┆           ┆           ┆           │
│       ┆            ┆ Brewery    ┆            ┆   ┆ Brewing   ┆           ┆           ┆           │
└───────┴────────────┴────────────┴────────────┴───┴───────────┴───────────┴───────────┴───────────┘

The dataset has the following structure

| Column Name | Description |
|-------------|-------------|
| `id_ba` / `id_rb` | Unique identifier for each brewery in BA/RB | 
| `name_ba` / `name_rb` | Name of the brewery in BA/RB |
| `location_ba` / `location_rb` | Geographic location of the brewery in BA/RB |
| `nbr_beers_ba` / `nbr_beers_rb` | Number of beers produced by the brewery in BA/RB |
| `diff_scores` | Difference in scores between BA and RB |
| `sim_scores` | Similarity in scores between BA and RB |

In [7]:
describe(df_breweries)

+--------------+---------+------------------+---------------+-----------+-----------------------+-------------------+
| Column       | Type    |   Not null count |   Nulls count |   Nulls % |   Unique values count |   Unique values % |
|--------------+---------+------------------+---------------+-----------+-----------------------+-------------------|
| id_ba        | Int64   |             8281 |             0 |    0.00 % |                  8281 |          100.00 % |
| location_ba  | String  |             8281 |             0 |    0.00 % |                   205 |            2.48 % |
| name_ba      | String  |             8281 |             0 |    0.00 % |                  8281 |          100.00 % |
| nbr_beers_ba | Int64   |             8281 |             0 |    0.00 % |                   231 |            2.79 % |
| id_rb        | Int64   |             8281 |             0 |    0.00 % |                  8235 |           99.44 % |
| location_rb  | String  |             8281 |           

In [8]:
describe(df_breweries, filters=["id_ba", "id_rb"])

+--------------+---------+------------------+---------------+-----------+-----------------------+-------------------+
| Column       | Type    |   Not null count |   Nulls count |   Nulls % |   Unique values count |   Unique values % |
|--------------+---------+------------------+---------------+-----------+-----------------------+-------------------|
| location_ba  | String  |             8281 |             0 |    0.00 % |                   205 |            2.48 % |
| name_ba      | String  |             8281 |             0 |    0.00 % |                  8281 |          100.00 % |
| nbr_beers_ba | Int64   |             8281 |             0 |    0.00 % |                   231 |            2.79 % |
| location_rb  | String  |             8281 |             0 |    0.00 % |                   205 |            2.48 % |
| name_rb      | String  |             8281 |             0 |    0.00 % |                  8235 |           99.44 % |
| nbr_beers_rb | Int64   |             8281 |           

For our breweries on both platforms, we have no missing values (in this matched data).

##### Users dataset

**Note**: The users_approx data is essentailly exactly the same however rather than users being matched exactly by username across the two platforms, they are matched using a TF-IDF vectorizer and cosine similarity. Users are matched if username is close enough and same location.

For each of our users (for each platform), we have their number of ratings/reviews, their ID, their name, the timestamp of when they joined and their location. The location is interesting for us, along with the number of reviews of each user.

In [9]:
df_users.sample(5)

shape: (5, 14)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ joined_ba ┆ location_ ┆ nbr_ratin ┆ nbr_revie ┆ … ┆ user_id_r ┆ user_name ┆ user_name ┆ sim_scor │
│ ---       ┆ ba        ┆ gs_ba     ┆ ws_ba     ┆   ┆ b         ┆ _rb       ┆ _lower_rb ┆ es       │
│ f64       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---      │
│           ┆ str       ┆ i64       ┆ i64       ┆   ┆ i64       ┆ str       ┆ str       ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 1.2521e9  ┆ United    ┆ 309       ┆ 0         ┆ … ┆ 2133      ┆ frogfan99 ┆ frogfan99 ┆ 0.866025 │
│           ┆ States,   ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ Texas     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 1.4902e9  ┆ Spain     ┆ 1         ┆ 0         ┆ … ┆ 252277    ┆ Ramonetti ┆ ramonetti ┆ 0.866025 │
│ 1.3000e9  ┆ United    ┆ 84        ┆ 84        ┆ … ┆ 116932    ┆ raszputin ┆ raszputin ┆ 1.0      │
│           ┆ States,   ┆           ┆           ┆   ┆           ┆ i         ┆ i         ┆          │
│           ┆ North     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ Carolina  ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 1.1201e9  ┆ United    ┆ 7         ┆ 7         ┆ … ┆ 33493     ┆ kevinphis ┆ kevinphis ┆ 1.0      │
│           ┆ States,   ┆           ┆           ┆   ┆           ┆ h         ┆ h         ┆          │
│           ┆ Oregon    ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 1.3930e9  ┆ United    ┆ 1920      ┆ 77        ┆ … ┆ 355050    ┆ domvan    ┆ domvan    ┆ 1.0      │
│           ┆ States,   ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│           ┆ Texas     ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

The dataset has the following structure
| Column Name | Description |
|-------------|-------------|
| `user_id_ba` / `user_id_rb` | Unique identifier for each user in BA/RB |
| `user_name_ba` / `user_name_rb` | Username of the reviewer in BA/RB |
| `user_name_lower_ba` / `user_name_lower_rb` | Lowercase username in BA/RB |
| `joined_ba` / `joined_rb` | Date when the user joined BA/RB |
| `location_ba` / `location_rb` | Geographic location of the user in BA/RB |
| `nbr_ratings_ba` / `nbr_ratings_rb` | Number of ratings submitted by the user in BA/RB |
| `nbr_reviews_ba` | Number of reviews submitted by the user in BA |

In [10]:
describe(df_users)

+--------------------+---------+------------------+---------------+-----------+-----------------------+-------------------+
| Column             | Type    |   Not null count |   Nulls count |   Nulls % |   Unique values count |   Unique values % |
|--------------------+---------+------------------+---------------+-----------+-----------------------+-------------------|
| joined_ba          | Float64 |             3341 |             0 |    0.00 % |                  2352 |           70.40 % |
| location_ba        | String  |             3341 |             0 |    0.00 % |                   111 |            3.32 % |
| nbr_ratings_ba     | Int64   |             3341 |             0 |    0.00 % |                   660 |           19.75 % |
| nbr_reviews_ba     | Int64   |             3341 |             0 |    0.00 % |                   475 |           14.22 % |
| user_id_ba         | String  |             3341 |             0 |    0.00 % |                  3316 |           99.25 % |
| user_n

In [11]:
describe_numbers(df_users, filters=["user_id_rb", "joined_ba", "joined_rb"])

+----------------+----------+-----------+-------+-------+-------+----------+-------+
| Column         |     Mean |       Std |   25% |   50% |   75% |      Min |   Max |
|----------------+----------+-----------+-------+-------+-------+----------+-------|
| nbr_ratings_ba |  211.612 |   685.867 |     2 |    13 |   106 |        1 | 12046 |
| nbr_reviews_ba |  118.232 |   454.174 |     1 |     3 |    29 |        0 |  7593 |
| nbr_ratings_rb |  270.923 |   1143.09 |     2 |     5 |    38 |        1 | 20678 |
| sim_scores     | 0.986855 | 0.0425066 |     1 |     1 |     1 | 0.800641 |     1 |
+----------------+----------+-----------+-------+-------+-------+----------+-------+


To make the `joined` columns more readable we cast them to a datetime object.

In [12]:
df_users = df_users.with_columns((pl.col("joined_ba").cast(pl.Int64) * 1000).cast(pl.Datetime("ms")))
df_users = df_users.with_columns((pl.col("joined_rb").cast(pl.Int64) * 1000).cast(pl.Datetime("ms")))
df_users.sample(5)

shape: (5, 14)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ joined_ba ┆ location_ ┆ nbr_ratin ┆ nbr_revie ┆ … ┆ user_id_r ┆ user_name ┆ user_name ┆ sim_scor │
│ ---       ┆ ba        ┆ gs_ba     ┆ ws_ba     ┆   ┆ b         ┆ _rb       ┆ _lower_rb ┆ es       │
│ datetime[ ┆ ---       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---      │
│ ms]       ┆ str       ┆ i64       ┆ i64       ┆   ┆ i64       ┆ str       ┆ str       ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 2013-08-0 ┆ United    ┆ 73        ┆ 2         ┆ … ┆ 109636    ┆ HoppyToda ┆ hoppytoda ┆ 1.0      │
│ 1         ┆ States,   ┆           ┆           ┆   ┆           ┆ y         ┆ y         ┆          │
│ 10:00:00  ┆ Oregon    ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2013-07-0 ┆ England   ┆ 11        ┆ 11        ┆ … ┆ 124233    ┆ EdKing    ┆ edking    ┆ 1.0      │
│ 1         ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 10:00:00  ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2010-05-1 ┆ United    ┆ 32        ┆ 32        ┆ … ┆ 106412    ┆ Konadrink ┆ konadrink ┆ 1.0      │
│ 0         ┆ States,   ┆           ┆           ┆   ┆           ┆ er        ┆ er        ┆          │
│ 10:00:00  ┆ Florida   ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2013-05-1 ┆ United    ┆ 1         ┆ 1         ┆ … ┆ 289453    ┆ darb85    ┆ darb85    ┆ 1.0      │
│ 1         ┆ States,   ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 10:00:00  ┆ Michigan  ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2008-12-0 ┆ United    ┆ 10        ┆ 10        ┆ … ┆ 81776     ┆ sonicstyl ┆ sonicstyl ┆ 1.0      │
│ 8         ┆ States,   ┆           ┆           ┆   ┆           ┆ ee        ┆ ee        ┆          │
│ 11:00:00  ┆ New York  ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

##### Ratings

In [13]:
df_ratings.sample(5)

shape: (5, 33)
┌────────┬────────────┬──────────┬────────────┬───┬──────────┬────────────┬────────────┬───────────┐
│ abv_ba ┆ appearance ┆ aroma_ba ┆ beer_id_ba ┆ … ┆ taste_rb ┆ text_rb    ┆ user_id_rb ┆ user_name │
│ ---    ┆ _ba        ┆ ---      ┆ ---        ┆   ┆ ---      ┆ ---        ┆ ---        ┆ _rb       │
│ f64    ┆ ---        ┆ f64      ┆ i64        ┆   ┆ f64      ┆ str        ┆ i64        ┆ ---       │
│        ┆ f64        ┆          ┆            ┆   ┆          ┆            ┆            ┆ str       │
╞════════╪════════════╪══════════╪════════════╪═══╪══════════╪════════════╪════════════╪═══════════╡
│ 6.0    ┆ 4.0        ┆ 4.0      ┆ 212102     ┆ … ┆ 8.0      ┆ 330 ml     ┆ 105791     ┆ superspak │
│        ┆            ┆          ┆            ┆   ┆          ┆ bottle     ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ into tulip ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ glass…     ┆            ┆           │
│ 6.3    ┆ 4.0        ┆ 3.75     ┆ 131281     ┆ … ┆ 7.0      ┆ This pours ┆ 147999     ┆ Lemke10   │
│        ┆            ┆          ┆            ┆   ┆          ┆ a dark     ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ amber      ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ color …    ┆            ┆           │
│ 6.2    ┆ 3.5        ┆ 3.5      ┆ 52475      ┆ … ┆ 6.0      ┆ Received   ┆ 60182      ┆ johnnnnie │
│        ┆            ┆          ┆            ┆   ┆          ┆ in trade   ┆            ┆ e         │
│        ┆            ┆          ┆            ┆   ┆          ┆ from       ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ mdascha…   ┆            ┆           │
│ 6.1    ┆ 4.0        ┆ 4.5      ┆ 22790      ┆ … ┆ 9.0      ┆ Notes from ┆ 70429      ┆ cakanator │
│        ┆            ┆          ┆            ┆   ┆          ┆ last year. ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ Bottle da… ┆            ┆           │
│ 5.3    ┆ 3.5        ┆ 4.0      ┆ 8698       ┆ … ┆ 6.0      ┆ Some       ┆ 11905      ┆ Sammy     │
│        ┆            ┆          ┆            ┆   ┆          ┆ citrus     ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ orange     ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ taste, but ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ …          ┆            ┆           │
└────────┴────────────┴──────────┴────────────┴───┴──────────┴────────────┴────────────┴───────────┘

In [14]:
describe(df_ratings)

+-----------------+---------+------------------+---------------+-----------+-----------------------+-------------------+
| Column          | Type    |   Not null count |   Nulls count |   Nulls % |   Unique values count |   Unique values % |
|-----------------+---------+------------------+---------------+-----------+-----------------------+-------------------|
| abv_ba          | Float64 |            21964 |             0 |    0.00 % |                   187 |            0.85 % |
| appearance_ba   | Float64 |            19295 |          2669 |   12.15 % |                    18 |            0.09 % |
| aroma_ba        | Float64 |            19295 |          2669 |   12.15 % |                    18 |            0.09 % |
| beer_id_ba      | Int64   |            21964 |             0 |    0.00 % |                  9025 |           41.09 % |
| beer_name_ba    | String  |            21964 |             0 |    0.00 % |                  8719 |           39.70 % |
| brewery_id_ba   | Int64   |   

The dataset has the following structure

| Column Name | Description |
|-------------|-------------|
| `beer_id_ba` / `beer_id_rb` | Unique identifier for the beer in BA/RB |
| `beer_name_ba` / `beer_name_rb` | Name of the beer in BA/RB |
| `brewery_id_ba` / `brewery_id_rb` | Unique identifier for the brewery in BA/RB |
| `brewery_name_ba` / `brewery_name_rb` | Name of the brewery in BA/RB |
| `style_ba` / `style_rb` | Beer style in BA/RB |
| `abv_ba` / `abv_rb` | Alcohol By Volume percentage in BA/RB |
| `user_id_ba` / `user_id_rb` | Unique identifier for the user in BA/RB |
| `user_name_ba` / `user_name_rb` | Username of the reviewer in BA/RB |
| `date_ba` / `date_rb` | Date of the review in BA/RB |
| `rating_ba` / `rating_rb` | Overall rating given by the user in BA/RB |
| `overall_ba` / `overall_rb` | Overall score in BA/RB |
| `appearance_ba` / `appearance_rb` | Appearance score in BA/RB |
| `aroma_ba` / `aroma_rb` | Aroma score in BA/RB |
| `palate_ba` / `palate_rb` | Palate score in BA/RB |
| `taste_ba` / `taste_rb` | Taste score in BA/RB |
| `text_ba` / `text_rb` | Review text in BA/RB |
| `review_ba` | Whether or not there is a BA review available |

In [15]:
describe_numbers(df_ratings, filters=["beer_id_ba", "beer_id_rb", "brewery_id_ba", "brewery_id_rb", "user_id_ba", "user_id_rb"])

+---------------+-------------+-------------+-------------+-------------+-------------+-------------+-------------+
| Column        |        Mean |         Std |         25% |         50% |         75% |         Min |         Max |
|---------------+-------------+-------------+-------------+-------------+-------------+-------------+-------------|
| abv_ba        |     6.80022 |     2.09619 |         5.2 |         6.4 |           8 |         0.5 |          39 |
| appearance_ba |     3.78329 |    0.562042 |         3.5 |           4 |           4 |           1 |           5 |
| aroma_ba      |     3.70404 |    0.597862 |         3.5 |        3.75 |           4 |           1 |           5 |
| date_ba       | 1.34367e+09 | 1.12069e+08 | 1.27098e+09 | 1.36835e+09 | 1.43471e+09 | 9.84136e+08 | 1.50141e+09 |
| overall_ba    |     3.73419 |    0.614882 |         3.5 |           4 |           4 |           1 |           5 |
| palate_ba     |     3.67871 |    0.618717 |         3.5 |        3.75 

Here we see:

1. Alcohol By Volume (ABV)
   - The `abv_ba` and `abv_rb` columns are identical again of course.

2. Rating Categories on BeerAdvocate (BA)
   - `appearance_ba`, `aroma_ba`, `palate_ba`, `taste_ba`, and `overall_ba` ratings on BeerAdvocate have similar means (ranging from 3.68 to 3.78) with a consistent standard deviation of around 0.6.
   - The median for each category is close to 4, suggesting that BeerAdvocate users tend to give positive ratings overall, as the maximum for these is 5.
   - The overall rating (`rating_ba`) has a mean of 3.71, indicating a generally favorable tendency in ratings across these categories.

3. Rating Categories on RateBeer (RB)
   - `appearance_rb`, `palate_rb`, and `rating_rb` have means similar to those on BeerAdvocate (3.53 to 3.69), suggesting that for these categories, user ratings are comparable across platforms.
   - `aroma_rb` and `taste_rb`, however, have higher maximum scores of 10, with means of 6.96 and 7.01, respectively. This difference likely reflects a different scoring system or metric on RateBeer, which will need to be accounted for in our analysis.
   - `overall_rb` shows a much broader scale, with a mean of 14.40 and a maximum of 20, indicating that RateBeer may use a larger rating scale for this category than BeerAdvocate does. Again this will need to be accounted for.

4. General Trends
   - BeerAdvocate ratings are generally more constrained within a 1-5 range for all categories, while RateBeer uses a broader scale for certain categories (`aroma_rb`, `taste_rb`, and `overall_rb`). This suggests a difference in rating systems that should be considered in cross-platform comparisons.
   - Despite these scale differences, the means for comparable categories like `appearance`, `palate`, and `rating` are fairly aligned between platforms, implying similar user perceptions on these aspects of beer.

Finally we'll convert the date into a datetime object.


In [16]:
df_ratings = df_ratings.with_columns((pl.col("date_ba").cast(pl.Int64) * 1000).cast(pl.Datetime("ms")))
df_ratings = df_ratings.with_columns((pl.col("date_rb").cast(pl.Int64) * 1000).cast(pl.Datetime("ms")))

df_ratings.sample(5)

shape: (5, 33)
┌────────┬────────────┬──────────┬────────────┬───┬──────────┬────────────┬────────────┬───────────┐
│ abv_ba ┆ appearance ┆ aroma_ba ┆ beer_id_ba ┆ … ┆ taste_rb ┆ text_rb    ┆ user_id_rb ┆ user_name │
│ ---    ┆ _ba        ┆ ---      ┆ ---        ┆   ┆ ---      ┆ ---        ┆ ---        ┆ _rb       │
│ f64    ┆ ---        ┆ f64      ┆ i64        ┆   ┆ f64      ┆ str        ┆ i64        ┆ ---       │
│        ┆ f64        ┆          ┆            ┆   ┆          ┆            ┆            ┆ str       │
╞════════╪════════════╪══════════╪════════════╪═══╪══════════╪════════════╪════════════╪═══════════╡
│ 4.5    ┆ 4.0        ┆ 3.5      ┆ 80811      ┆ … ┆ 8.0      ┆ 12 oz      ┆ 137945     ┆ rangerred │
│        ┆            ┆          ┆            ┆   ┆          ┆ bottle     ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ into a     ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ pint       ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ glass…     ┆            ┆           │
│ 5.0    ┆ 3.75       ┆ 3.25     ┆ 87019      ┆ … ┆ 6.0      ┆ Pours pale ┆ 69847      ┆ popery    │
│        ┆            ┆          ┆            ┆   ┆          ┆ gold and   ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ mildly     ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ haz…       ┆            ┆           │
│ 7.5    ┆ 4.0        ┆ 4.0      ┆ 238194     ┆ … ┆ 8.0      ┆ 500 ml     ┆ 105791     ┆ superspak │
│        ┆            ┆          ┆            ┆   ┆          ┆ bottle     ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ into tulip ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ glass…     ┆            ┆           │
│ 6.1    ┆ 3.5        ┆ 3.5      ┆ 61889      ┆ … ┆ 6.0      ┆ Appearance ┆ 32031      ┆ ChainGang │
│        ┆            ┆          ┆            ┆   ┆          ┆ :          ┆            ┆ Guy       │
│        ┆            ┆          ┆            ┆   ┆          ┆ Dispensed  ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ from the…  ┆            ┆           │
│ 4.7    ┆ 4.0        ┆ 4.0      ┆ 52096      ┆ … ┆ 6.0      ┆ On cask at ┆ 96987      ┆ Huhzubend │
│        ┆            ┆          ┆            ┆   ┆          ┆ Barcade in ┆            ┆ ah        │
│        ┆            ┆          ┆            ┆   ┆          ┆ Brooklyn…  ┆            ┆           │
└────────┴────────────┴──────────┴────────────┴───┴──────────┴────────────┴────────────┴───────────┘

## Processing and saving the cleaned data
Here we'll sync the data of the matched beer dataset with the one of the RateBeer and BeerAdvocate. This is needed because we've removed some data from the BeerAdvocate and RateBeer datasets during the data cleaning process and we need to ensure that the matched beer dataset is in sync with the other two datasets.
##### Preliminary

In [17]:
BEERADVOCATE_PATH = '../../../data/BeerAdvocate/processed'
RATEBEER_PATH = '../../../data/RateBeer/processed'


##### Users dataset

In [18]:
df_filtered_users = df_users.select(["user_id_ba", "user_id_rb", "sim_scores"])
print("Data before cleaning", df_filtered_users.shape[0])

# BeerAdvocate
df_users_ba = pl.read_parquet(BEERADVOCATE_PATH + "/users.pq")
df_users_ba = df_users_ba.rename({col: f"{col}_ba" for col in df_users_ba.columns})
df_filtered_users = df_filtered_users.join(df_users_ba, on="user_id_ba", how="inner")
del df_users_ba

# RateBeer
df_users_rb = pl.read_parquet(RATEBEER_PATH + "/users.pq")
df_users_rb = df_users_rb.rename({col: f"{col}_rb" for col in df_users_rb.columns})
df_filtered_users = df_filtered_users.join(df_users_rb, on="user_id_rb", how="inner")
del df_users_rb

print("Data after cleaning", df_filtered_users.shape[0])
print(f"Percentage of data lost: {100 - (df_filtered_users.shape[0] / df_users.shape[0] * 100)}%")

# Save the data
df_filtered_users.write_parquet(DST_DATA_PATH + "/users.pq")

Data before cleaning 3341
Data after cleaning 3341
Percentage of data lost: 0.0%


In [19]:
df_filtered_users.sample(5)

shape: (5, 15)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ user_id_b ┆ user_id_r ┆ sim_score ┆ user_name ┆ … ┆ location_ ┆ nbr_ratin ┆ nbr_revie ┆ nbr_inte │
│ a         ┆ b         ┆ s         ┆ _ba       ┆   ┆ rb        ┆ gs_rb     ┆ ws_rb     ┆ ractions │
│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ _rb      │
│ str       ┆ i64       ┆ f64       ┆ str       ┆   ┆ str       ┆ i32       ┆ i32       ┆ ---      │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ i32      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ mcnealc31 ┆ 119458    ┆ 1.0       ┆ McNealc31 ┆ … ┆ United    ┆ 211       ┆ 211       ┆ 211      │
│ .335631   ┆           ┆           ┆           ┆   ┆ States,   ┆           ┆           ┆          │
│           ┆           ┆           ┆           ┆   ┆ Ohio      ┆           ┆           ┆          │
│ holybrew. ┆ 50421     ┆ 1.0       ┆ holybrew  ┆ … ┆ India     ┆ 14        ┆ 14        ┆ 14       │
│ 185690    ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ doob.1172 ┆ 73490     ┆ 1.0       ┆ Doob      ┆ … ┆ England   ┆ 1         ┆ 1         ┆ 1        │
│ 376       ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ metalgdog ┆ 127030    ┆ 1.0       ┆ metalgdog ┆ … ┆ United    ┆ 1         ┆ 1         ┆ 1        │
│ .540033   ┆           ┆           ┆           ┆   ┆ States,   ┆           ┆           ┆          │
│           ┆           ┆           ┆           ┆   ┆ New York  ┆           ┆           ┆          │
│ deekyn.64 ┆ 137131    ┆ 1.0       ┆ deekyn    ┆ … ┆ United    ┆ 1135      ┆ 1135      ┆ 1135     │
│ 3035      ┆           ┆           ┆           ┆   ┆ States,   ┆           ┆           ┆          │
│           ┆           ┆           ┆           ┆   ┆ Californi ┆           ┆           ┆          │
│           ┆           ┆           ┆           ┆   ┆ a         ┆           ┆           ┆          │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

##### Beers dataset

In [20]:
df_filtered_beers = df_beers.select(["beer_id_ba", "beer_id_rb", "diff_scores", "sim_scores"])
print("Data before cleaning", df_filtered_beers.shape[0])

# BeerAdvocate
df_beers_ba = pl.read_parquet(BEERADVOCATE_PATH + "/beers.pq")
df_beers_ba = df_beers_ba.rename({col: f"{col}_ba" for col in df_beers_ba.columns})
df_filtered_beers = df_filtered_beers.join(df_beers_ba, on="beer_id_ba", how="inner")

# RateBeer
df_beers_rb = pl.read_parquet(RATEBEER_PATH + "/beers.pq")
df_beers_rb = df_beers_rb.rename({col: f"{col}_rb" for col in df_beers_rb.columns})
df_filtered_beers = df_filtered_beers.join(df_beers_rb, on="beer_id_rb", how="inner")

print("Data after cleaning", df_filtered_beers.shape[0])
print(f"Percentage of data lost: {(1 - df_filtered_beers.shape[0] / df_beers.shape[0]) * 100:.2f}%")

# Save the data
df_filtered_beers.write_parquet(DST_DATA_PATH + "/beers.pq")

Data before cleaning 45640
Data after cleaning 45636
Percentage of data lost: 0.01%


In [21]:
df_filtered_beers.sample(5)

shape: (5, 26)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ beer_id_b ┆ beer_id_r ┆ diff_scor ┆ sim_score ┆ … ┆ rating_sc ┆ nbr_inter ┆ nbr_ratin ┆ nbr_revi │
│ a         ┆ b         ┆ es        ┆ s         ┆   ┆ ore_media ┆ actions_r ┆ gs_rb     ┆ ews_rb   │
│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ n_rb      ┆ b         ┆ ---       ┆ ---      │
│ i64       ┆ i64       ┆ f64       ┆ f64       ┆   ┆ ---       ┆ ---       ┆ i64       ┆ i64      │
│           ┆           ┆           ┆           ┆   ┆ f64       ┆ i64       ┆           ┆          │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 118056    ┆ 258812    ┆ 1.0       ┆ 1.0       ┆ … ┆ 3.4       ┆ 7         ┆ 7         ┆ 7        │
│ 88753     ┆ 25982     ┆ 0.69937   ┆ 1.0       ┆ … ┆ 3.0       ┆ 97        ┆ 97        ┆ 97       │
│ 122583    ┆ 268000    ┆ 1.0       ┆ 1.0       ┆ … ┆ 3.8       ┆ 33        ┆ 33        ┆ 33       │
│ 247157    ┆ 446460    ┆ 1.0       ┆ 1.0       ┆ … ┆ 3.8       ┆ 31        ┆ 31        ┆ 31       │
│ 40664     ┆ 83368     ┆ 1.0       ┆ 1.0       ┆ … ┆ 3.3       ┆ 1         ┆ 1         ┆ 1        │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

##### Breweries dataset

In [22]:
df_filtered_breweries = df_breweries.select(["id_ba", "id_rb", "diff_scores", "sim_scores"])
print("Data before cleaning", df_filtered_breweries.shape[0])

# BeerAdvocate
df_breweries_ba = pl.read_parquet(BEERADVOCATE_PATH + "/breweries.pq")
df_breweries_ba = df_breweries_ba.rename({col: f"{col}_ba" for col in df_breweries_ba.columns})
df_filtered_breweries = df_filtered_breweries.join(df_breweries_ba, on="id_ba", how="inner")

# RateBeer
df_breweries_rb = pl.read_parquet(RATEBEER_PATH + "/breweries.pq")
df_breweries_rb = df_breweries_rb.rename({col: f"{col}_rb" for col in df_breweries_rb.columns})
df_filtered_breweries = df_filtered_breweries.join(df_breweries_rb, on="id_rb", how="inner")

print("Data after cleaning", df_filtered_breweries.shape[0])
print(f"Percentage of data lost: {(1 - df_filtered_breweries.shape[0] / df_breweries.shape[0]) * 100:.2f}%")

# Save the data
df_filtered_breweries.write_parquet(DST_DATA_PATH + "/breweries.pq")

Data before cleaning 8281
Data after cleaning 7428
Percentage of data lost: 10.30%


In [23]:
df_filtered_breweries.sample(5)

shape: (5, 10)
┌───────┬───────┬─────────────┬────────────┬───┬────────────┬────────────┬────────────┬────────────┐
│ id_ba ┆ id_rb ┆ diff_scores ┆ sim_scores ┆ … ┆ beers_coun ┆ location_r ┆ name_rb    ┆ beers_coun │
│ ---   ┆ ---   ┆ ---         ┆ ---        ┆   ┆ t_ba       ┆ b          ┆ ---        ┆ t_rb       │
│ i64   ┆ i64   ┆ f64         ┆ f64        ┆   ┆ ---        ┆ ---        ┆ str        ┆ ---        │
│       ┆       ┆             ┆            ┆   ┆ i64        ┆ str        ┆            ┆ i32        │
╞═══════╪═══════╪═════════════╪════════════╪═══╪════════════╪════════════╪════════════╪════════════╡
│ 41074 ┆ 22687 ┆ 0.342566    ┆ 1.0        ┆ … ┆ 10         ┆ Australia  ┆ Pirate     ┆ 18         │
│       ┆       ┆             ┆            ┆   ┆            ┆            ┆ Life       ┆            │
│       ┆       ┆             ┆            ┆   ┆            ┆            ┆ Brewing    ┆            │
│ 20914 ┆ 1342  ┆ 0.845173    ┆ 0.845173   ┆ … ┆ 37         ┆ Lithuania  ┆ Volfas     ┆ 94         │
│       ┆       ┆             ┆            ┆   ┆            ┆            ┆ Engelman   ┆            │
│       ┆       ┆             ┆            ┆   ┆            ┆            ┆ (Olvi)     ┆            │
│ 40399 ┆ 16413 ┆ 1.0         ┆ 1.0        ┆ … ┆ 1          ┆ Spain      ┆ Gothia     ┆ 3          │
│       ┆       ┆             ┆            ┆   ┆            ┆            ┆ Launia     ┆            │
│ 41596 ┆ 29499 ┆ 0.770562    ┆ 1.0        ┆ … ┆ 13         ┆ United     ┆ Helicon    ┆ 23         │
│       ┆       ┆             ┆            ┆   ┆            ┆ States,    ┆ Brewing    ┆            │
│       ┆       ┆             ┆            ┆   ┆            ┆ Pennsylvan ┆            ┆            │
│       ┆       ┆             ┆            ┆   ┆            ┆ ia         ┆            ┆            │
│ 37557 ┆ 19205 ┆ 0.69256     ┆ 1.0        ┆ … ┆ 5          ┆ Switzerlan ┆ La         ┆ 29         │
│       ┆       ┆             ┆            ┆   ┆            ┆ d          ┆ Nébuleuse  ┆            │
└───────┴───────┴─────────────┴────────────┴───┴────────────┴────────────┴────────────┴────────────┘

##### Ratings dataset

In [24]:
df_ratings.head()

shape: (5, 33)
┌────────┬────────────┬──────────┬────────────┬───┬──────────┬────────────┬────────────┬───────────┐
│ abv_ba ┆ appearance ┆ aroma_ba ┆ beer_id_ba ┆ … ┆ taste_rb ┆ text_rb    ┆ user_id_rb ┆ user_name │
│ ---    ┆ _ba        ┆ ---      ┆ ---        ┆   ┆ ---      ┆ ---        ┆ ---        ┆ _rb       │
│ f64    ┆ ---        ┆ f64      ┆ i64        ┆   ┆ f64      ┆ str        ┆ i64        ┆ ---       │
│        ┆ f64        ┆          ┆            ┆   ┆          ┆            ┆            ┆ str       │
╞════════╪════════════╪══════════╪════════════╪═══╪══════════╪════════════╪════════════╪═══════════╡
│ 11.3   ┆ 4.5        ┆ 4.5      ┆ 645        ┆ … ┆ 9.0      ┆ a)  Geruch ┆ 83106      ┆ Erzengel  │
│        ┆            ┆          ┆            ┆   ┆          ┆ malzig-sch ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ wer-süÅ  ┆            ┆           │
│ 5.0    ┆ null       ┆ null     ┆ 28191      ┆ … ┆ 4.0      ┆ Can. Weak  ┆ 91324      ┆ visionthi │
│        ┆            ┆          ┆            ┆   ┆          ┆ and        ┆            ┆ ng        │
│        ┆            ┆          ┆            ┆   ┆          ┆ watery,    ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ not the …  ┆            ┆           │
│ 5.0    ┆ 3.5        ┆ 3.5      ┆ 57911      ┆ … ┆ 8.0      ┆ Bottle @   ┆ 98624      ┆ tiong     │
│        ┆            ┆          ┆            ┆   ┆          ┆ One Pint   ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ Pub,       ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ Helsink…   ┆            ┆           │
│ 5.0    ┆ 4.0        ┆ 3.5      ┆ 57913      ┆ … ┆ 9.0      ┆ Draught    ┆ 98624      ┆ tiong     │
│        ┆            ┆          ┆            ┆   ┆          ┆ @ Pikkuli ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ ntu,       ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ Helsink…   ┆            ┆           │
│ 6.0    ┆ 4.0        ┆ 4.0      ┆ 81125      ┆ … ┆ 8.0      ┆ 750ml bott ┆ 98624      ┆ tiong     │
│        ┆            ┆          ┆            ┆   ┆          ┆ leBottling ┆            ┆           │
│        ┆            ┆          ┆            ┆   ┆          ┆ date: 201… ┆            ┆           │
└────────┴────────────┴──────────┴────────────┴───┴──────────┴────────────┴────────────┴───────────┘

In [25]:
df_ratings_filtered = df_ratings.select(["beer_id_ba", "brewery_id_ba", "user_id_ba", "beer_id_rb", "brewery_id_rb", "user_id_rb"])

print("Data before cleaning", df_ratings_filtered.shape[0])

# BeerAdvocate
df_ratings_ba = pl.read_parquet(BEERADVOCATE_PATH + "/ratings.pq")
df_ratings_ba = df_ratings_ba.rename({col: f"{col}_ba" for col in df_ratings_ba.columns})
df_ratings_filtered = df_ratings_filtered.join(df_ratings_ba, on=["beer_id_ba", "brewery_id_ba", "user_id_ba"], how="inner")

# RateBeer
df_ratings_rb = pl.read_parquet(RATEBEER_PATH + "/ratings.pq")
df_ratings_rb = df_ratings_rb.rename({col: f"{col}_rb" for col in df_ratings_rb.columns})
df_ratings_filtered = df_ratings_filtered.join(df_ratings_rb, on=["beer_id_rb", "brewery_id_rb", "user_id_rb"], how="inner")

print("Data after cleaning", df_ratings_filtered.shape[0])
print(f"Percentage of data lost: {(1 - df_ratings_filtered.shape[0] / df_ratings.shape[0]) * 100:.2f}%")

# Save the data
df_ratings_filtered.write_parquet(DST_DATA_PATH + "/ratings.pq")

Data before cleaning 21964
Data after cleaning 21964
Percentage of data lost: 0.00%


### Discussion
As we observed in the data exploration of RateBeer and MatchedBeer we don't lose lots of information regarding beers, users or ratings and the only relevant data lost is for the breweries (where we has lost approximately 10% of the data). Given that no review has been lost, and the filtering has been done in the processing of the files used to do the matching here, we can be confident that the data lost weren't relevant given that they didn't have any reviews.